lib/hanami/configuration.rb

```diff
@@ -3,268 +3,176 @@
 require "uri"
 require "concurrent/hash"
 require "concurrent/array"
+require "dry/configurable"
 require "dry/inflector"
 require "pathname"
 require "zeitwerk"
 
+require_relative "application/settings/loader"
+require_relative "configuration/logger"
+require_relative "configuration/middleware"
+require_relative "configuration/router"
+require_relative "configuration/sessions"
+
 module Hanami
   # Hanami application configuration
   #
   # @since 2.0.0
   #
   # rubocop:disable Metrics/ClassLength
   class Configuration
-    require_relative "configuration/middleware"
-    require_relative "configuration/router"
-    require_relative "configuration/sessions"
+    include Dry::Configurable
+
+    DEFAULT_ENVIRONMENTS = Concurrent::Hash.new { |h, k| h[k] = Concurrent::Array.new }
+    private_constant :DEFAULT_ENVIRONMENTS
 
     attr_reader :actions
+    attr_reader :middleware
+    attr_reader :router
     attr_reader :views
 
-    # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
-    def initialize(env:)
-      @settings = Concurrent::Hash.new
-
-      self.autoloader = Zeitwerk::Loader.new
+    attr_reader :environments
+    private :environments
 
-      self.env = env
-      self.environments = DEFAULT_ENVIRONMENTS.clone
+    def initialize(env:)
+      @environments = DEFAULT_ENVIRONMENTS.clone
+      config.env = env
 
+      # Some default setting values must be assigned at initialize-time to ensure they
+      # have appropriate values for the current application
       self.root = Dir.pwd
-      self.slices_dir = DEFAULT_SLICES_DIR
-      settings[:slices] = {}
-
-      self.settings_path = DEFAULT_SETTINGS_PATH
-      self.settings_loader_options = {}
-
-      self.base_url = DEFAULT_BASE_URL
-
-      self.logger   = DEFAULT_LOGGER.clone
-      self.rack_logger_filter_params = DEFAULT_RACK_LOGGER_FILTER_PARAMS.clone
-      self.sessions = DEFAULT_SESSIONS
-
-      self.router     = Router.new(base_url)
-      self.middleware = Middleware.new
-
-      self.inflections = Dry::Inflector.new
+      self.autoloader = Zeitwerk::Loader.new
 
+      # Config for actions (same for views, below) may not be available if the gem isn't
+      # loaded; fall back to a null config object if it's missing
       @actions = begin
         require_path = "hanami/action/application_configuration"
         require require_path
         Hanami::Action::ApplicationConfiguration.new
       rescue LoadError => e
         raise e unless e.path == require_path
-        Object.new
+        require_relative "configuration/null_configuration"
+        NullConfiguration.new
       end
 
+      @middleware = Middleware.new
+
+      @router = Router.new(self)
+
       @views = begin
         require_path = "hanami/view/application_configuration"
         require require_path
         Hanami::View::ApplicationConfiguration.new
```

Comment:
I actually wonder whether we should make these `setting` declarations like most of the others... by having a mix of both settings as well as directly-assigned attributes/ivars, I feel we lose a bit of consistency (and these latter attributes don't show up when inspecting the configuration's `settings` etc.)

That said, in these cases, there's no need for e.g. the `middleware` or `router` objects to be wholly assigned or _replaced_ like the other settings, so perhaps just leaving them as directly-assigned attributes makes the most since, since we can avoid having writer methods for them.